src/legacy_util.cpp

```diff
@@ -109,6 +109,74 @@ copy_fd(int fd_in, int fd_out, bool fd_in_is_file)
   return true;
 }
 
+// Copy all data from fd_in to fd_out, omitting any ANSI CSI escape sequences
+// whose action character is in the specified set.
+bool
+copy_fd_strip_csi_seqs(int fd_in, int fd_out, const char* strip_actions)
```

Comment:
> This was of course hard for you to foresee...

Indeed. I could have saved myself quite a lot of work if I had thrown in the towel and operated on a single memory buffer from the get-go. It was some job to write that function to operate on an indefinite-length stream of data while minimizing syscall overhead and still functioning correctly in the case that an ANSI CSI sequence straddles a read boundary.

> Would it make sense to convert this function to a function with the signature `std::string strip_color_codes(const std::string& data)` or so instead?

I can certainly do that, although I'd recommend that the parameter type be `std::string_view` instead of `const std::string&` in case the caller doesn't have a `std::string` object on hand but does have a contiguous span of character data.

> Also, please consider moving the function to `Utils.cpp` and make it ~~as C++-y you can~~ a bit more C++-y if possible.

I had considered defining it as a template function that takes as an argument a function object that is called back with the locations of any ANSI CSI sequences found in the input text and that is allowed to modify those sequences (typically by just clearing them, but alternatively it could rewrite them), but I didn't like the performance implications of what I would have to do if the callback function object happened to rewrite a CSI sequence to something *longer* than the original sequence, which could cause the output text segment to become larger than the input text segment and thus no longer fit in the fixed-size buffer on the stack.

Just how C++-y am I allowed to be? I routinely write C++2a code in my own projects, and I'm often having to polyfill libstdc++ because, even as of the latest GCC release, it's still not up to snuff with what's coming out of the standards committee. I would assume you don't want ccache to require GCC 10. So, what C++ version are you targeting?

Reply:
@jrosdahl: See my [commit](https://github.com/ccache/ccache/pull/596/commits/3ee05833c366f47dbf51a99ad4b2bca13242de18) that changes the CSI sequence processing to operate entirely in memory and makes it significantly more C++-y.

Reply:
> I could have saved myself quite a lot of work 

Sorry to hear that.

> I can certainly do that, although I'd recommend that the parameter type be std::string_view instead of const std::string& in case the caller doesn't have a std::string object on hand but does have a contiguous span of character data.

Sure, I was only trying to illustrate the principle.

> Just how C++-y am I allowed to be?

C++11, as can be read under [_Code style_ in CONTRIBUTING.md](https://github.com/ccache/ccache/blob/master/CONTRIBUTING.md#code-style).

Also, my comment was a bit tongue-in-cheek. I just wanted to hint that you don't need to imimtate the C-style code in legacy_utils.cpp.

> See my commit that changes the CSI sequence processing to operate entirely in memory and makes it significantly more C++-y.

Thanks.